src/jenkins.py

```diff
@@ -105,3 +200,169 @@ def unlock_jenkins(connectable_container: Container) -> None:
     version = get_version()
     connectable_container.push(LAST_EXEC_VERSION_PATH, version, encoding="utf-8", make_dirs=True)
     connectable_container.push(WIZARD_VERSION_PATH, version, encoding="utf-8", make_dirs=True)
+
+
+def _install_config(connectable_container: ops.Container, jnlp_port: str) -> None:
+    """Install jenkins-config.xml.
+
+    Args:
+        connectable_container: The connectable Jenkins workload container.
+        jnlp_port: The JNLP port to communicate with the agents.
+    """
+    env = jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True)
+    template = env.get_template("jenkins-config.xml.j2")
+    config = template.render(jnlp_port=jnlp_port)
+    connectable_container.push(CONFIG_FILE_PATH, config)
+
+
+def _install_plugins(connectable_container: ops.Container) -> None:
+    """Install Jenkins plugins.
+
+    Download Jenkins plugins. A restart is required for the changes to take effect.
+
+    Args:
+        connectable_container: The connectable Jenkins workload container.
+
+    Raises:
+        JenkinsPluginError: if an error occurred installing the plugin.
+    """
+    plugins = " ".join(set(REQUIRED_PLUGINS))
+    proc: ops.pebble.ExecProcess = connectable_container.exec(
+        [
+            "java",
+            "-jar",
+            "jenkins-plugin-manager-2.12.11.jar",
+            "-w",
+            "jenkins.war",
+            "-d",
+            str(PLUGINS_PATH),
+            "-p",
+            plugins,
+        ],
+        working_dir="/srv/jenkins/",
+        timeout=300,
+    )
+    try:
+        proc.wait_output()
+    except (ops.pebble.ChangeError, ops.pebble.ExecError) as exc:
+        logger.error("Failed to install plugins, %s", exc)
+        raise JenkinsPluginError("Failed to install plugins.") from exc
+
+
+def bootstrap(
+    connectable_container: ops.Container,
+    jnlp_port: str,
+) -> None:
+    """Initialize and install Jenkins.
+
+    Args:
+        connectable_container: The connectable Jenkins workload container.
+        jnlp_port: The JNLP port to communicate with the agents.
+
+    Raises:
+        JenkinsBootstrapError: if there was an error installing given plugins or required plugins.
+    """
+    _unlock_wizard(connectable_container)
+    _install_config(connectable_container, jnlp_port)
+    try:
+        _install_plugins(connectable_container)
+    except JenkinsPluginError as exc:
+        raise JenkinsBootstrapError("Failed to bootstrap Jenkins.") from exc
+
+
+def is_boostrapped(connectable_container: ops.Container) -> bool:
```

Comment:
I'm wondering if we should cache this as a property, since once it's bootstrapped there are few chances that this would change

Reply:
There's a plan to remove this function in favour of doing a `pebble.restart` :D